src/GraphQL/Types/Collections/SchemaTypes.cs

```diff
@@ -646,11 +647,62 @@ private void AddTypeIfNotRegistered(IGraphType type, TypeCollectionContext conte
             var (namedType, namedType2) = type.GetNamedTypes();
             namedType ??= context.ResolveType(namedType2!);
 
-            var foundType = this[namedType.Name];
-            if (foundType == null)
+            var existingType = this[namedType.Name];
+            if (existingType is null)
             {
                 AddType(namedType, context);
             }
+            else
+            {
+                EnsureTypeEquality(existingType, namedType);
+            }
+        }
+
+        private void EnsureTypeEquality(IGraphType existingType, IGraphType newType)
+        {
+            if (ReferenceEquals(existingType, newType))
+            {
+                return;
+            }
+
+            if (existingType.GetType() != newType.GetType())
+            {
+                throw new InvalidOperationException($"Type '{existingType.GetType().Name}' is already registered but being re-registered using different type '{newType.GetType().Name}'");
```

Comment:
I suggest we expand this error message a bit to include the GraphQL type name that is registered a second time.